test-common/utils.c

```diff
@@ -698,107 +877,136 @@ generate_delta_files (GFile *repo,
+                       &checksum,
+                       error))
+    return FALSE;
 
-      if (commit_number > 0)
-        {
-          /* O(N^2), sadly */
-          const OstreeCollectionRef *last_ref = get_last_ref (subserver->ref_to_commit,
-                                                              commit_number);
-          /* Get the checksum of the commit on the last ref, since
-           * it may have changed in the meantime */
-          if (!get_current_commit_checksum (subserver->repo,
-                                            (last_ref != NULL) ? last_ref : collection_ref,
-                                            &old_checksum,
-                                            error))
-            return FALSE;
-        }
+  if (parent_commit_info != NULL)
+    {
+      const gchar *old_checksum = g_hash_table_lookup (subserver->commits_in_repo,
+                                                       GUINT_TO_POINTER (parent_commit_info->sequence_number));
 
-      if (!prepare_commit (subserver->repo,
-                           subserver->tree,
-                           commit_number,
-                           collection_ref,
-                           subserver->gpg_home,
-                           subserver->keyid,
-                           subserver->additional_directories_for_commit,
-                           subserver->additional_files_for_commit,
-                           subserver->additional_metadata_for_commit,
-                           &checksum,
-                           error))
-        return FALSE;
+      g_assert_nonnull (old_checksum);
 
-      if (commit_number > 0)
-        {
-          if (!generate_delta_files (subserver->repo,
-                                     old_checksum,
-                                     checksum,
-                                     error))
-            return FALSE;
-        }
+      if (!generate_delta_files (subserver->repo,
+                                 old_checksum,
+                                 checksum,
+                                 error))
+        return FALSE;
     }
 
+  /* Insert commit checksum into hashtable */
+  g_hash_table_insert (subserver->commits_in_repo,
+                       GUINT_TO_POINTER (commit_info->sequence_number),
+                       g_strdup (checksum));
+
+  return TRUE;
+}
+
+/* Updates the subserver to reflect the state of
+ * the internal commit graph.  This involves
+ * creating the commits, generating ref files and
+ * delta files, and updating the summary.
+ */
+static gboolean
+update_commits (EosTestSubserver *subserver,
+                GError **error)
+{
+  g_auto(CmdResult) cmd = CMD_RESULT_CLEARED;
+
+  if (!eos_test_updater_commit_graph_walk (subserver->commit_graph,
+                                           make_commit_if_not_available,
+                                           subserver,
+                                           error))
+    return FALSE;
```

Comment:
This could also be implemented through the use of a generator a-la `GHashTableIter`, though I didn't think it was worth the extra complexity.

Reply:
Probably not worth it.